File: 백준/Gold/15686. 치킨 배달/치킨 배달.js
const fs = require('fs');
const input = fs.readFileSync('/dev/stdin').toString().trim().split('\n');

const [N, M] = input.shift().split(' ').map(Number);
const board = input.map((row) => row.split(' ').map(Number));
const chicken = [];
const home = [];
let min = Number.MAX_SAFE_INTEGER;

for (let i = 0; i < N; i++) {
    for (let j = 0; j < N; j++) {
        if (board[i][j] === 2) chicken.push([i, j]);
        else if (board[i][j] === 1) home.push([i, j]);
    }
}

function getDist(arr) {
    let sum = 0;
    for (let i = 0; i < home.length; i++) {
        let min = Number.MAX_SAFE_INTEGER;
        for (let j of arr) {
            min = Math.min(min, Math.abs(home[i][0] - chicken[j][0]) + Math.abs(home[i][1] - chicken[j][1]));
        }

        sum += min;
    }
    return sum;
}

function dfs(target, now, arr, depth) {
    if (depth === target) {
        min = Math.min(min, getDist(arr));
        return;
    }

    for (let i = now; i < chicken.length; i++) {
        arr.push(i);
        dfs(target, i + 1, arr, depth + 1);
        arr.pop();
    }
}

for (let i = 1; i <= M; i++) {
    dfs(i, 0, [], 0);
}

console.log(min);

File: 백준/Gold/15686. 치킨 배달/치킨 배달.py
from itertools import combinations

N, M = map(int, input().split())
board = [list(map(int, input().split())) for _ in range(N)]

chicken = []
for y in range(N):
    for x in range(N):
        if board[y][x] == 2:
            chicken.append([y,x])

def get_length(comb):
    total = 0
    for y in range(N):
        for x in range(N):
            if board[y][x] == 1:
                sum = int(1e9)
                for k in comb:
                    sum = min(abs(y - k[0]) + abs(x - k[1]), sum)

                total += sum

    return total




answer = int(1e9)
for i in range(1, M+1):
    for x in combinations(chicken, i):
        answer = min(get_length(x), answer)

print(answer)